DOCSP-24549: delete fundamentals (#23)
mongodb/docs-csharp: source/fundamentals/crud/write-operations/delete.txt

.. _csharp-delete-guide:

=================

> Delete a Document

Reviewer: issue: The title doesn't exactly match the page content. I expected it to only contain information about how to delete *a* document. Perhaps simply "Delete Documents"?

Author: Title was taken from the corresponding GoDoc page, but I agree, the title should be more accurate even if it means changing the task oriented language

=================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to remove documents from your MongoDB
collections using delete operations.

Sample Data
~~~~~~~~~~~

The example in this guide use the ``restaurants`` collection
from the ``sample_restaurants`` database. The documents in this
collection are modeled by the following ``Restaurant`` class:

.. literalinclude:: /includes/fundamentals/code-examples/crud/delete.cs
   :language: csharp
   :dedent:
   :start-after: start-model
   :end-before: end-model

This collection is from the :atlas:`sample datasets </sample-data>` provided
by Atlas. See the :ref:`<csharp-quickstart>` to learn how to create a free MongoDB cluster
and load this sample data.

Delete Operations
-----------------

> Use **delete operations** to remove data from MongoDB. Delete operations
> use the ``DeleteOne()`` method, which deletes *the first document* that
> matches your query filter, and the ``DeleteMany()`` method, which
> deletes *all documents* that match the filter.

Reviewer: issue: This puts the relevant information further from the reader by structure.
issue: Since delete operations aren't defined (nor do they likely need defined), do we need to bold the term?

```suggestion
Use delete operations to remove documents that match a **query filter**. <define query filter>. Perform delete operations in MongoDB with the following methods:

- ``DeleteOne()`` method, which deletes *the first document* that
matches your query filter
- ``DeleteMany()`` method, which
deletes *all documents* that match a query filter
```

Author: migrated info from Parameters section to this section.

> Delete a Document

Reviewer: ```suggestion
Delete One Document
```

~~~~~~~~~~~~~~~~~

> The following code shows how you can use the asynchronous

Reviewer: ```suggestion
The following code shows how to use the asynchronous
```

``DeleteOneAsync()`` method or the synchronous ``DeleteOne()`` method to
delete one document.

.. tabs::

   .. tab:: Asynchronous
      :tabid: delete-one-async

      .. code-block:: csharp
         :copyable: true

         var result = await _restaurantsCollection.DeleteOneAsync(filter);

   .. tab:: Synchronous
      :tabid: delete-one-sync

      .. code-block:: csharp
         :copyable: true

         var result = _restaurantsCollection.DeleteOne(filter);

Delete Multiple Documents
~~~~~~~~~~~~~~~~~~~~~~~~~

> The following code shows how you can use the asynchronous

Reviewer: ```suggestion
The following code shows how to use the asynchronous
```

``DeleteManyAsync()`` method or the synchronous ``DeleteMany()`` method to
delete all matched documents.

.. tabs::

   .. tab:: Asynchronous
      :tabid: delete-many-async

      .. code-block:: csharp
         :copyable: true

         var result = await _restaurantsCollection.DeleteManyAsync(filter);

   .. tab:: Synchronous
      :tabid: delete-many-sync

      .. code-block:: csharp
         :copyable: true

         var result = _restaurantsCollection.DeleteMany(filter);

Find runnable examples using these methods under :ref:`additional

> information <additional-info>`.

Reviewer: issue: I wonder if this should be highlighted as a tip?

Parameters
~~~~~~~~~~

> The ``DeleteOne()`` and ``DeleteMany()`` methods expect you to pass a

Reviewer: ```suggestion
The ``DeleteOne()`` and ``DeleteMany()`` methods require you to pass a
```

*query filter* specifying which documents to match. The query filter
determines which records are selected for deletion based on the criteria

> in :manual:`the query filter document</core/document/#query-filter-documents>`.

Reviewer: issue: The definition for query filter follows its use up above

Author: moved content to introduction

Both methods optionally take a ``DeleteOptions`` type as an additional parameter,
which represents options you can use to configure the delete operation.
If you don't specify any ``DeleteOptions`` properties, the driver does
not customize the delete operation.

The ``DeleteOptions`` type allows you to configure options with the
following properties:

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Property
     - Description

   * - ``Collation``
     - | Gets or sets the type of language collation to use when sorting results.

   * - ``Comment``
     - | Gets or sets the comment for the operation.

   * - ``Hint``
     - | Gets or sets the index to use to scan for documents.

   * - ``Let``

>      - | Gets or sets the let document.

Reviewer: issue: some of these might we worth linking to additional resources, especially collation and let

Example
~~~~~~~

The following code uses the ``DeleteMany()`` method to search on the

> "borough_1" index and delete all documents whose ``address.street``

Reviewer: issue: this struck me as a bit of personification
```suggestion
"borough_1" index and delete all documents where the ``address.street``
```

field value includes the phrase "Pearl Street":

.. io-code-block::
   :copyable: true

   .. input::
      :language: csharp

      var filter = Builders<Restaurant>.Filter
          .Regex("address.street", "Pearl Street");

      DeleteOptions opts = new DeleteOptions { Hint = "borough_1" };

      WriteLine("Deleting documents...");
      var result = _restaurantsCollection.DeleteMany(filter, opts);
      
      WriteLine($"Deleted documents: {result.DeletedCount}");

>       WriteLine($"Result acknowledged? {result.IsAcknowledged}");

Reviewer: issue: can this be moved to source from a formatted .cs file?

Author: according to the [wiki page for the IO block](https://wiki.corp.mongodb.com/display/DE/How-To%3A+Add+Code+Output+for+Code+Snippets), I can pass a filepath to the [delete.cs file](https://github.com/mongodb/docs-csharp/blob/98d2f01d54a257e5fd826bef620486a0c32ac7df/source/includes/fundamentals/code-examples/crud/delete.cs) that I did create, but it doesn't seem like I can select the lines to include like I can with a literalinclude. So I just pasted the relevant code directly into the block. Let me know if you think I should just create a .cs file with just this code to include.

Reviewer: I think it's okay in this instance since it's so short.
